### ISSUE TITLE
在某一个step模型的输出突然全变为!，kl变为Nan

### ISSUE THREAD

**Issue by yushuiwx** [volcengine/verl#751]
same with 

- https://github.com/volcengine/verl/issues/747
- https://github.com/volcengine/verl/issues/721

注意到我们的模型使用的deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B，!对应的token id是0.

![Image](https://github.com/user-attachments/assets/65b746e5-a24f-4db9-ab3b-6ec28eb09c11)

**Comment by ftgreat** [comment]
请问跑起来至少需要多少卡，谢谢

**Comment by AIBionics** [comment]
可以提供对应的训练脚本和数据源，这样方便项目人员复现和排查

**Comment by yushuiwx** [comment]
> Could you please try the following steps:
> 
> * Resume the last checkpoint and make sure the nan persists
> * Resume from the last checkpoint and step save frequency to 1
> * Resume from the last checkpoint and make sure nan immediately appears in the following step
> * Add a breakpoint in the ppo_actor loss function and print old_log_probs, log_probs and ratio to find out which part in the loss computation causes nan.

Hello, after careful analysis, I found that the issue lies in the calculation of `low_var_kl`. In the problematic step, the loss has normal values (without NaN or Inf), but during backpropagation, the `exp` operation in `ratio = torch.exp(kl)` within `low_var_kl` （https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386）causes extreme values, leading to NaN in `kl`. My temporary solution is to use `torch.clamp(kl, min=-5, max=5)` to constrain `kl` before computing `ratio`. This approach works, but I feel uncertain about it. Could you kindly help confirm whether this is an appropriate solution?

**Comment by yushuiwx** [comment]
> Could you please try the following steps:
> 
> * Resume the last checkpoint and make sure the nan persists
> * Resume from the last checkpoint and step save frequency to 1
> * Resume from the last checkpoint and make sure nan immediately appears in the following step
> * Add a breakpoint in the ppo_actor loss function and print old_log_probs, log_probs and ratio to find out which part in the loss computation causes nan.

I found several similar issues [#747,](https://github.com/volcengine/verl/issues/747), [727](https://github.com/volcengine/verl/issues/727), etc . I suspect that they might be caused by the same underlying problem.

**Comment by LYMDLUT** [comment]
Welcome to read and quote our article. We have proven that in GRPO, the k2 loss is an analytical expression, which is better than the k3 loss. We have also provided additional correction methods for off-policy and std in GRPO.  
[k2 loss就是比k3 loss好，以及grpo off-policy、clip_std()]https://zhuanlan.zhihu.com/p/1892008158626546312

@misc{Liu2025KL,
author       = {YiMing Liu},
title        = {Rethinking KL Divergence in RLHF: From Single Sample to Mini-Batch to Expectation},
year         = {2025},
url          = {https://www.notion.so/Rethinking-KL-Divergence-in-RLHF-From-Single-Sample-to-Mini-Batch-to-Expectation-1c18637cdeb3800ab47cd01d3fa33ea5},
note         = {Notion Blog}
}

**Comment by chuangzhidan** [comment]
> Welcome to read and quote our article. We have proven that in GRPO, the k2 loss is an analytical expression, which is better than the k3 loss. When the model generates a low probability, the gradient of the k3 loss will experience exponential growth, causing the model to collapse. However, the k2 loss does not have this problem. We have also provided additional correction methods for off-policy and std in GRPO. [k2 loss就是比k3 loss好，以及grpo off-policy、clip_std()]https://zhuanlan.zhihu.com/p/1892008158626546312
> 
> [@misc](https://github.com/misc){Liu2025KL, author = {YiMing Liu}, title = {Rethinking KL Divergence in RLHF: From Single Sample to Mini-Batch to Expectation}, year = {2025}, url = {[https://www.notion.so/Rethinking-KL-Divergence-in-RLHF-From-Single-Sample-to-Mini-Batch-to-Expectation-1c18637cdeb3800ab47cd01d3fa33ea5}](https://www.notion.so/Rethinking-KL-Divergence-in-RLHF-From-Single-Sample-to-Mini-Batch-to-Expectation-1c18637cdeb3800ab47cd01d3fa33ea5%7D), note = {Notion Blog} }

how to use k2 loss in verl

**Comment by chuangzhidan** [comment]
> > Could you please try the following steps:
> > 
> > * Resume the last checkpoint and make sure the nan persists
> > * Resume from the last checkpoint and step save frequency to 1
> > * Resume from the last checkpoint and make sure nan immediately appears in the following step
> > * Add a breakpoint in the ppo_actor loss function and print old_log_probs, log_probs and ratio to find out which part in the loss computation causes nan.
> 
> Hello, after careful analysis, I found that the issue lies in the calculation of `low_var_kl`. In the problematic step, the loss has normal values (without NaN or Inf), but during backpropagation, the `exp` operation in `ratio = torch.exp(kl)` within `low_var_kl` （[https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386）causes](https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386%EF%BC%89causes) extreme values, leading to NaN in `kl`. My temporary solution is to use `torch.clamp(kl, min=-5, max=5)` to constrain `kl` before computing `ratio`. **This approach works, but I feel uncertain about it. Could you kindly help confirm whether this is an appropriate solution?**

current code in main branch is this,is this ok?
    if kl_penalty == 'low_var_kl':
        kl = ref_logprob - logprob
        ratio = torch.exp(kl)
        kld = (ratio - kl - 1).contiguous()
        return torch.clamp(kld, min=-10, max=10)

**Comment by LYMDLUT** [comment]
> > > Could you please try the following steps:
> > > 
> > > * Resume the last checkpoint and make sure the nan persists
> > > * Resume from the last checkpoint and step save frequency to 1
> > > * Resume from the last checkpoint and make sure nan immediately appears in the following step
> > > * Add a breakpoint in the ppo_actor loss function and print old_log_probs, log_probs and ratio to find out which part in the loss computation causes nan.
> > 
> > 
> > Hello, after careful analysis, I found that the issue lies in the calculation of `low_var_kl`. In the problematic step, the loss has normal values (without NaN or Inf), but during backpropagation, the `exp` operation in `ratio = torch.exp(kl)` within `low_var_kl` （[https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386）causes](https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386%EF%BC%89causes) extreme values, leading to NaN in `kl`. My temporary solution is to use `torch.clamp(kl, min=-5, max=5)` to constrain `kl` before computing `ratio`. **This approach works, but I feel uncertain about it. Could you kindly help confirm whether this is an appropriate solution?**
> 
> current code in main branch is this,is this ok? if kl_penalty == 'low_var_kl': kl = ref_logprob - logprob ratio = torch.exp(kl) kld = (ratio - kl - 1).contiguous() return torch.clamp(kld, min=-10, max=10)

kl_penalty == "mse"

**Comment by chuangzhidan** [comment]
> > > > Could you please try the following steps:
> > > > 
> > > > * Resume the last checkpoint and make sure the nan persists
> > > > * Resume from the last checkpoint and step save frequency to 1
> > > > * Resume from the last checkpoint and make sure nan immediately appears in the following step
> > > > * Add a breakpoint in the ppo_actor loss function and print old_log_probs, log_probs and ratio to find out which part in the loss computation causes nan.
> > > 
> > > 
> > > Hello, after careful analysis, I found that the issue lies in the calculation of `low_var_kl`. In the problematic step, the loss has normal values (without NaN or Inf), but during backpropagation, the `exp` operation in `ratio = torch.exp(kl)` within `low_var_kl` （[https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386）causes](https://github.com/volcengine/verl/blob/8cae42dc29736d0802ded43c5ecf67a809d56bd8/verl/trainer/ppo/core_algos.py#L386%EF%BC%89causes) extreme values, leading to NaN in `kl`. My temporary solution is to use `torch.clamp(kl, min=-5, max=5)` to constrain `kl` before computing `ratio`. **This approach works, but I feel uncertain about it. Could you kindly help confirm whether this is an appropriate solution?**
> > 
> > 
> > current code in main branch is this,is this ok? if kl_penalty == 'low_var_kl': kl = ref_logprob - logprob ratio = torch.exp(kl) kld = (ratio - kl - 1).contiguous() return torch.clamp(kld, min=-10, max=10)
> 
> kl_penalty == "mse"
> 
> If it is effective after the modification, please let me know.

works for reinforece++ and grpo ? 只需要改参数设置取值，不改代码？我试下

**Comment by Xia723** [comment]
都试过了还是不行，有什么最新的解决方案么

**Comment by koalazf99** [comment]
Same issue here. we set `kl_coeff = 0`, but still failed.

**Comment by takagi97** [comment]
我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。
具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置
        actor_rollout_ref.rollout.enforce_eager=False \
        actor_rollout_ref.rollout.free_cache_engine=False \
就可以了

**Comment by zihaolucky** [comment]
> 我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。 具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置 actor_rollout_ref.rollout.enforce_eager=False actor_rollout_ref.rollout.free_cache_engine=False 就可以了

This doesn't work for me. :(

**Comment by qilong-zhang** [comment]
> > 我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。 具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置 actor_rollout_ref.rollout.enforce_eager=False actor_rollout_ref.rollout.free_cache_engine=False 就可以了
> 
> This doesn't work for me. :(

me too

**Comment by zhenghaoxu-gatech** [comment]
similar problem, outputting '!' even no metric is NaN

**Comment by Yux1angJi** [comment]
> 我用vllm 0.8.2 + 2节点跑的时候也遇到这个问题，经人提醒，将vllm的v1引擎关掉之后，一切正常了。 具体操作是，设置export VLLM_USE_V1=0，并且训练参数里面不要使用设置 actor_rollout_ref.rollout.enforce_eager=False actor_rollout_ref.rollout.free_cache_engine=False 就可以了

@takagi97 请问是指要把actor_rollout_ref.rollout.enforce_eager=False actor_rollout_ref.rollout.free_cache_engine=False这两个参数设为False，还是不设置保留默认的True呢

**Comment by xjtupy** [comment]
请问解决了吗

**Comment by Kirrito-k423** [comment]
中文：目前不建议把这个 issue 当作已解决关闭。#789 只是在 `grad_norm` 非有限时跳过更新，属于防止异常权重更新的保护措施，并不能证明 `!` 输出 / KL NaN 的根因已经修复。相关 #891 仍是 open，评论里也有用户反馈 clamp、关闭 KL 或关闭 vLLM V1 并不总是有效。

当前 main 相对 2025 年 3-4 月已有一些相关变化：`low_var_kl/k3` 会在 `exp` 前 clamp `ref_logprob - logprob`，文档也支持 `low_var_kl+` / `k3+`，用于保持 k3 前向值但用 k2 梯度做 straight-through。建议先在最新 main 或最新 release 上确认是否仍复现，并说明使用的是 `low_var_kl`、`low_var_kl+`、`mse/k2` 还是关闭 actor KL loss。

继续定位需要基于当前代码的最小复现：完整启动命令/配置、verl commit 或版本、vLLM 版本、是否设置 `VLLM_USE_V1`、rollout 后端和 mode、模型与可共享的小数据样本；同时请贴出出问题 step 前后 `old_log_probs`、`log_probs`、`ref_logprob`、`actor/kl_loss`、`actor/ppo_kl`、`actor/entropy_loss`、`actor/grad_norm` 的非有限值检查结果。

English: I would not treat this issue as resolved yet. #789 only skips the optimizer update when `grad_norm` is non-finite; it is a guard against applying a bad update, not evidence that the root cause of `!` outputs / KL NaN is fixed. The related #891 is still open, and several comments report that clamp, disabling KL, or disabling vLLM V1 are not universal fixes.

Current main has changed since the March-April 2025 reports: `low_var_kl/k3` now clamps `ref_logprob - logprob` before `exp`, and the docs support `low_var_kl+` / `k3+`, which keeps the k3 forward value while using the k2 gradient via a straight-through trick. Please first check whether the problem still reproduces on latest main or latest release, and state whether the run uses `low_var_kl`, `low_var_kl+`, `mse/k2`, or disables actor KL loss.

To continue debugging, we need a minimal reproduction on current code: full launch command/config, verl commit or version, vLLM version, whether `VLLM_USE_V1` is set, rollout backend and mode, model plus a shareable small data sample. Please also include finite/NaN checks around the first failing step for `old_log_probs`, `log_probs`, `ref_logprob`, `actor/kl_loss`, `actor/ppo_kl`, `actor/entropy_loss`, and `actor/grad_norm`.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
